### ISSUE TITLE
Is it possible to prevent api/auth/session from being called multiple times?

### ISSUE THREAD

**Issue by frontlearner28** [nextauthjs/next-auth#6630]
### Question 💬

Currently, I use getSession() in my Axios interceptor, to control valid session. The problem is that the session is being called multiple times on pages with many API calls, increasing my page load time too much.

Is there any way to prevent this?

![CleanShot 2023-02-06 at 11 46 25](https://user-images.githubusercontent.com/124594929/217002775-cbe528a5-5b30-4db4-a976-ec6eca2de577.png)

### How to reproduce ☕️

To reproduce the problem, use getSession in an Axios interceptor and execute many calls on one page

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by balazsorban44** [comment]
You should not use `useSession` for axios, as it's a React api. check out the `getSession` method instead.

**Comment by frontlearner28** [comment]
@balazsorban44 but in my question I explained that I use getSession, at no point did I comment about useSession. Can you please reopen the issue? :c

**Comment by balazsorban44** [comment]
I need my glasses... 👀 Could you attach a reproduction in any case? this is likely not a NextAuth.js big for what it's worth, but how you use axios, I assume.

**Comment by Ali-hd** [comment]
Hi @frontlearner28 I am facing the same issue as you, did you find out have you can reduce the api requests every time you use getSession() ?

**Comment by Ali-hd** [comment]
> I need my glasses... 👀 This is likely still not a NextAuth.js bug for what it's worth, but how you use axios, I assume.
> 
> `getSession` is essentially just a json request to the `/api/auth/session` endpoint, it does no polling or the like.

Hi @balazsorban44 I understand that getSession is only used to get session info but is there a better way to use it? In client side I need to send the access token with many requests. With every request I need to call the method getSession() to get the Auth token, I don't have any other way in mind. Can I save the session data locally in the browser or something?

**Comment by patelmilanun** [comment]
@Ali-hd did u found any solution to this. @balazsorban44 can u suggest some better approach or link to something that can help here?

**Comment by balazsorban44** [comment]
You can for example proxy the API requests through the Next.js backend and attach the session there. You can get the session server-side without extra network trips.

**Comment by patelmilanun** [comment]
Well actually, api/auth/session is basically the same. So from client side we are calling proxy api then it will be calling the api just like /api/auth/session as both of the things will be at next.js backend.

So either call api/auth/session and get token and then call the actual api or call the proxy api. I think in 2nd approch we can save 1 network trip from client side.

**Comment by Ali-hd** [comment]
> @Ali-hd did u found any solution to this. @balazsorban44 can u suggest some better approach or link to something that can help here?

what I did to avoid making a session network request before every api call is to basically just use the useSession() hook whenever I need to get the user token for my clientside requests instead of using the getSession()

avoid using getSession() unless you really need it. 

`const useAxiosAuth = () => {
  const { data: session, status } = useSession()
  const [loading, setLoading] = useState(false)
  const [isLoadingAuth, setIsLoadingAuth] = useState(true)
  const userToken = session?.user?.token ? `Bearer ${session.user.token}` : undefined

  useEffect(() => {
    if (session === undefined || status === "loading") {
      setIsLoadingAuth(true)
      return
    }
    setIsLoadingAuth(false)
  }, [session, status])

  const withAuth = useCallback(
    async (config: AxiosRequestConfig): Promise<HttpResponse> => {
      setLoading(true)
      const headers = { ...config?.headers, Authorization: userToken }

      try {
        const options = { ...config, headers }
        // you can use your axios interceptor instance instead
        let res = await axios.request(options)

        if (res.status === 401) {
          // refresh token with getSession() or do something else
          const newSession = await getSession()
          options.headers = { Authorization: `Bearer ${newSession?.user?.token}` }
          res = await axios.request(options)
          // sign out if it's failed again
          // ... signOut({ redirect: true, callbackUrl: "/" })
        }

        return res
      } catch (err) {
        // send error to sentry or something
        return {
          success: false,
          data: null,
        }
      } finally {
        setLoading(false)
      }
    },
    [userToken]
  )

  return {
    loading,
    isLoadingAuth,
    withAuth,
  }
}

export default useAxiosAuth`

And then I do this 

```
const { withAuth, isLoadingAuth, loading } = useAxiosAuth()
  
  useEffect(() => {
    // wait for session to load
    if (isLoadingAuth) return

    // you can add this to another file and import it here
    const axiosOptions= {
      headers: {},
      url: `/user/update`,
      method: "post",
      data: {},
    }

    const fetchData = async () => {
      const { success, data } = await withAuth(axiosOptions)
      // do something
    }
    fetchData()
  }, [router.isReady, withAuth, isLoadingAuth])
  ``` 
  
  Not the cleanest way but it gets the job done.  You can use the hook all over the app and it will only  call session once per window load

**Comment by patelmilanun** [comment]
TL;DR;
Use https://github.com/nextauthjs/next-auth-example/blob/main/app/client-example/page.tsx and set the ["zustand"](https://github.com/pmndrs/zustand?tab=readme-ov-file#readingwriting-state-and-reacting-to-changes-outside-of-components) state in client component so in https://github.com/nextauthjs/next-auth-example/blob/main/components/client-example.tsx

So actually I was implementing it for 2 things. First is openapi-fetch client to attach auth on every request similar to axios and is already implemented using the getSession() which was causing the problem and cant make it hook as it might break and will have to do refactor since it is not react component. Second is to check permission access for specific function of application where I need to compare permission required to access that resource vs the session permissions.

So I ended up using state management library "zustand" note that u should not use it to share state across server component and client component. So using their official document for how to implement auth in client component (basically call auth() and get session, then take that session and pass it along with session provider which is wrapping the client component, and use useSession()). I used this trick on main layout file and then just in client component I set the state of zustand. This allowed me to access latest value of zustand state everywhere even in non-component file.

So u can implement this approach. Thanks for help guys!

**Comment by nghianguyen119** [comment]
I saw this issue after migrating to V5, but if it's the issue of next-auth then I can rest assured, the team need time to refactor

**Comment by john-subba** [comment]
> I'm using V5 and this keeps happening.. session and jwt callbacks happening 6 times one after another :/ I tried everything

Yeah. I am rethinking whether to use this or not. The session callback api get's called so many times. Even if i have not used useSession() hook

**Comment by HiroakiLion** [comment]
Having the same issue, for now I am caching the data I received to prevent multiple api calls based on useSession, but everytime i use useSession, the session calls are increasing rapidly.

**Comment by john-subba** [comment]
> Having the same issue, for now I am caching the data I received to prevent multiple api calls based on useSession, but everytime i use useSession, the session calls are increasing rapidly.

i think if u use the session provider in specific components that u need u can tackle this issue even without caching it.
i previously wrapped everything with the provider in the layout.tsx maybe that was the reason it get's called multiple times.

**Comment by anshuopinion** [comment]
`import { getBackendServerBaseUrlByEnvironment } from "@/utils/environmentUtil";
import axios from "axios";
import { getSession } from "next-auth/react";

// Singleton instance of Axios Auth Client
let axiosAuthClientInstance: ReturnType<typeof axiosAuthClient> | null = null;

/**
 * Get Axios Auth Client Instance
 * @dev - critical wrapper for handling api authorization to the backend
 *
 * Returns the singleton instance of the Axios Auth Client. If the instance
 * does not exist, it creates one.
 *
 * @returns {AxiosInstance} The singleton Axios instance.
 */
export const getAxiosAuthClientInstance = () => {
  if (!axiosAuthClientInstance) {
    axiosAuthClientInstance = axiosAuthClient();
  }
  return axiosAuthClientInstance;
};

/**
 * Axios Auth Client
 *
 * Creates an Axios instance configured to include the authorization token
 * from the NextAuth session in the request headers. This instance is created
 * once and reused throughout the application.
 *
 * @returns {AxiosInstance} The configured Axios instance.
 */
const axiosAuthClient = () => {
  const instance = axios.create({
    baseURL: getBackendServerBaseUrlByEnvironment(),
    headers: {
      "Content-Type": "application/json",
    },
  });

  instance.interceptors.request.use(async (config) => {
    const session = await getSession();
    if (session && !config.headers["Authorization"]) {
      config.headers["Authorization"] = `Bearer ${session.accessToken}`;
    }
    return config;
  });

  instance.interceptors.response.use(
    (response) => response,
    (error) => {
      // Handle error response
      console.error("Request failed", error);
      return Promise.reject(error);
    }
  );

  return instance;
};

export default getAxiosAuthClientInstance;
`

try this solution

**Comment by anshuopinion** [comment]
```  import { getBackendServerBaseUrlByEnvironment } from "@/utils/environmentUtil";
import axios from "axios";
import { getSession } from "next-auth/react";

// Singleton instance of Axios Auth Client
let axiosAuthClientInstance: ReturnType<typeof axiosAuthClient> | null = null;

/**
 * Get Axios Auth Client Instance
 * @dev - critical wrapper for handling api authorization to the backend
 *
 * Returns the singleton instance of the Axios Auth Client. If the instance
 * does not exist, it creates one.
 *
 * @returns {AxiosInstance} The singleton Axios instance.
 */
export const getAxiosAuthClientInstance = () => {
  if (!axiosAuthClientInstance) {
    axiosAuthClientInstance = axiosAuthClient();
  }
  return axiosAuthClientInstance;
};

/**
 * Axios Auth Client
 *
 * Creates an Axios instance configured to include the authorization token
 * from the NextAuth session in the request headers. This instance is created
 * once and reused throughout the application.
 *
 * @returns {AxiosInstance} The configured Axios instance.
 */
const axiosAuthClient = () => {
  const instance = axios.create({
    baseURL: getBackendServerBaseUrlByEnvironment(),
    headers: {
      "Content-Type": "application/json",
    },
  });

  instance.interceptors.request.use(async (config) => {
    const session = await getSession();
    if (session && !config.headers["Authorization"]) {
      config.headers["Authorization"] = `Bearer ${session.accessToken}`;
    }
    return config;
  });

  instance.interceptors.response.use(
    (response) => response,
    (error) => {
      // Handle error response
      console.error("Request failed", error);
      return Promise.reject(error);
    }
  );

  return instance;
};

export default getAxiosAuthClientInstance;
 ```

Try this solution

**Comment by anshuopinion** [comment]
```
import { getBackendServerBaseUrlByEnvironment } from "@/utils/environmentUtil";
import axios from "axios";
import { getSession } from "next-auth/react";

// Singleton instance of Axios Auth Client
let axiosAuthClientInstance: ReturnType<typeof axiosAuthClient> | null = null;

/**
 * Get Axios Auth Client Instance
 * @dev - critical wrapper for handling api authorization to the backend
 *
 * Returns the singleton instance of the Axios Auth Client. If the instance
 * does not exist, it creates one.
 *
 * @returns {AxiosInstance} The singleton Axios instance.
 */
export const getAxiosAuthClientInstance = () => {
  if (!axiosAuthClientInstance) {
    axiosAuthClientInstance = axiosAuthClient();
  }
  return axiosAuthClientInstance;
};

/**
 * Axios Auth Client
 *
 * Creates an Axios instance configured to include the authorization token
 * from the NextAuth session in the request headers. This instance is created
 * once and reused throughout the application.
 *
 * @returns {AxiosInstance} The configured Axios instance.
 */
const axiosAuthClient = () => {
  const instance = axios.create({
    baseURL: getBackendServerBaseUrlByEnvironment(),
    headers: {
      "Content-Type": "application/json",
    },
  });

  instance.interceptors.request.use(async (config) => {
    const session = await getSession();
    if (session && !config.headers["Authorization"]) {
      config.headers["Authorization"] = `Bearer ${session.accessToken}`;
    }
    return config;
  });

  instance.interceptors.response.use(
    (response) => response,
    (error) => {
      // Handle error response
      console.error("Request failed", error);
      return Promise.reject(error);
    }
  );

  return instance;
};

export default getAxiosAuthClientInstance;

```

Try this solution

**Comment by lunarW1TCH** [comment]
If anyone stumbles upon it I solved it by fetching the session in main layout of my app and passing it to a component which sets it in a zustand store. I am not calling useSession anymore as it seems to be the culprit.

**Comment by john-subba** [comment]
> If anyone stumbles upon it I solved it by fetching the session in main layout of my app and passing it to a component which sets it in a zustand store. I am not calling useSession anymore as it seems to be the culprit.

useSession with the sessionProvider is the culprit if u use getServerSession it doesnt give the issue
but the problem if to use update() function to update the session u need to use useSession hook no other way around it.

**Comment by Renkas** [comment]
Why is this closed? It has not been solved ...

**Comment by iamvinny** [comment]
I have the following code to obtain the session in a server component:

```tsx
import { auth } from "@/auth"
import Form from "./_components/Form";

export default async function LoginPage() {
    const session = await auth()
    
    if(session) { 
        return (
            <div>{JSON.stringify(session)}</div>
        )
    }

    return (
        <Form />
    )

}
```

But for some reason, it is executing 7 times
![Screenshot 2024-07-18 at 5 50 19 PM](https://github.com/user-attachments/assets/6f28add9-bedd-435c-9f63-e0150ccefd7e)

These are my callbacks:
```tsx
    callbacks: {
        async jwt({ token, account, user }: any) {
            if (user) {
                token.user = user
                token.accessToken = user.access_token
            }
            return token
        },
        async session({ session, token }: any) {
            session.accessToken = token.access_token
            session.user = token.user
            return session
        },
    },
```

**Comment by waseemthefword** [comment]
Does any found a optimal way to do it ? without repeating session like multiple times.

**Comment by NoahGdev** [comment]
bump

**Comment by lakinmindfire** [comment]
Do we have a solution yet ?

**Comment by MagisterUnivers** [comment]
Jesus. That /session is on rampage now (14 times call on just watching the page, and switch tabs)

**Comment by MrMissx** [comment]
I resolve it when downgrading to `5.0.0-beta.20`

**Comment by michaeltroya** [comment]
looks like they decreased in `5.0.0-beta.20`, still getting 4 calls but better than the 11 I was getting on `beta.18`

**Comment by balazsorban44** [comment]
For anyone reading this, this is not a bug.

Depending on your code, it could happen for multiple reasons.

If anyone would like me to dissect their exact usecase a bit more and see if we can document gotchas better or fix any potential bugs, please open a new bug report with a reproduction. This issue has been closed, and we generally don't monitor these. The OP's question about axios should have been answered, I'm seeing diverse questions in follow-up comments now. 

Here is some of the general advice:

1. for `await auth()` calls, if you see too many of them, they are not cached per request by default, so naturally any logic to your session will execute as many times as `auth()` was called during render

Solution: https://react.dev/reference/react/cache

We are considering to add this by default

2. for `useSession`/ calls to `/api/auth/session`

- Make sure polling isn't enabled/configured correctly
- Multiple instances of `SessionProvider`'s might cause extra calls, usually you don't need more than one of these

With the App Router, we would however recommend relying on `auth()` and pass down session data to your client components on a need basis.


I hope this helps! 🙏

**Comment by nimbit-software** [comment]
@balazsorban44  could you expound on the react/cache solution please?

**Comment by Yegoroot** [comment]
```js
// CashedSessionContext.tsx
import { createContext, useContext, useState, useEffect, ReactNode } from 'react'
import { useSession } from 'next-auth/react'
import { Session } from 'next-auth'
import { setupAxiosInterceptors } from '@/api/utils/axios'

const CachedSessionContext = createContext<Session | null>(null)

interface SessionProviderProps {
  children: ReactNode
}

export function CachedSessionProvider({ children }: SessionProviderProps) {
  const { data: session } = useSession()
  const [cachedSession, setCachedSession] = useState<Session | null>(session)

  useEffect(() => {
    if (session) {
      setCachedSession(session)
      setupAxiosInterceptors(session) // HEEEERE
    }
  }, [session])

  return (
    <CachedSessionContext.Provider value={cachedSession}>
      {children}
    </CachedSessionContext.Provider>
  )
}


export function useCachedSession() {
  return useContext(CachedSessionContext)
}

```

axios instance
```
export const privateAxiosInstance = axios.create({
  baseURL: API_URL,
  withCredentials: true,
  headers: {
    'Content-Type': 'application/json',
  },
})


export function setupAxiosInterceptors(session: Session) {
  privateAxiosInstance.interceptors.request.use(
    (config) => {
      if (session?.accessToken) {
        config.headers.Authorization = `Bearer ${session.accessToken}`
      }
      return config
    },
    (error) => {
      const formattedError = {
        status: error.response?.status,
        message: error.response?.data?.message || error.message,
      }
      return Promise.reject(formattedError) 
    }
  )
}
```

now you can use `useCachedSession` and axios is configured

**Comment by amaryassa** [comment]
On  `"next-auth": "^5.0.0-beta.25"` , I also encountered this issue: every time the page gains focus, it triggers calls to the session. To work around this, I set `refetchOnWindowFocus={false}` to the <SessionProvider>  like this:

```tsx
 <SessionProvider session={session} refetchOnWindowFocus={false}>
        {children}
 </SessionProvider>
```

However, I haven’t fully explored whether this workaround could cause any side effects or regressions in the app.

We can also add an `refetchInterval`. Check out the [SessionProviderProps](https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/lib/client.ts#L106) for more details.

**Comment by jheredev** [comment]
Any solutions for this problem? Please help.

**Comment by tmfsol** [comment]
I just experimented with next-auth, I found it quite interesting, but also quite cumbersome, hehe.
I also use axios interceptor and also have the same problem, I think I basically need to store the session somewhere to avoid unnecessary callback.
Currently, my solution is to memories cache.

![code](https://github.com/user-attachments/assets/db22b9d2-94c3-49c0-a244-8f6523c0867f)
![2](https://github.com/user-attachments/assets/013785b4-8a82-4072-a893-a73dcae6ebd7)
![3](https://github.com/user-attachments/assets/fa5eaa73-6d8e-4f8c-8c9e-ab20b7b0748a)

**Comment by tmfsol** [comment]
My solution.
```
import { getSession as getSessionClient } from 'next-auth/react';

let cachedSession: any = null;

const makeSession = async () => {
	try {
		const session = isClient ? await getSessionClient() : await auth();

		return session || null;
	} catch {
		return null;
	}
};

export const getSession = async ({ cache } = { cache: false }) => {
	if (!cache) {
		return await makeSession();
	}

	cachedSession = cachedSession || makeSession();
	cachedSession = await cachedSession;

	if (cachedSession && Date.now() <= new Date(cachedSession.expires).getTime()) {
		return cachedSession;
	}

	cachedSession = await makeSession();

	return cachedSession;
};

export const attachJwtToRequest = async (request: AxiosRequestConfig) => {
	const session = await getSession({ cache: true });

	if (!session) {
		return;
	}

	if (request.headers) {
		if (request.url === 'auth/refresh-token') {
			request.headers['refresh-token'] = `Bearer ${session.refreshToken}`;
		} else {
			request.headers['Authorization'] = `Bearer ${session.accessToken}`;
		}
	}
};

**Comment by emanueldev1** [comment]
Hello everyone, for those who keep getting this problem I found a very simple solution based on several comments seen here and observing a little how the provider works, it is as easy as making your main layout asynchronous where you have the SessionProvider and make it look like the following example


```jsx
import { SessionProvider } from "next-auth/react";
import { auth } from "@/auth";


export default async function RootLayout({ children }) {
  const session = await auth(); // obtains the session only once on page loading
  return (
    <html lang="en">
      <body>  
          <SessionProvider session={session} refetchOnWindowFocus={false}> {/* disable the refresh on focus window and use the session object from the async function auth() one unique time*/}
              {children}
          </SessionProvider>
      </body>
    </html>
  );
}

```

With this I went from having between 6 and 11 calls to /api/auth/session to only one when the page loads and if I need it again I use the useSession or if it is on the server I still use the await auth().

**Comment by died14** [comment]
After looking for a solution for days, the solution of  @emanueldev1 finally solved the problem of receiving many requests to api/auth/session.

I had to make a slight modification because I couldn't use SessionProvider in a server side component.

Here is my implementation:

```typescript
// Layout.tsx
// Server side
import { getServerSession, Session } from "next-auth";

export default async function RootLayout({ children }: { children: React.ReactNode; }) {
	const session = await getServerSession() as Session;
	return (
		<html>
			<body>
				<AuthContext session={session}>
						{children}
				</AuthContext>
			</body>
		</html>
	);
}
``` 

```typescript
// AuthContext.tsx
'use client';

import { Session } from 'next-auth';
import { SessionProvider} from 'next-auth/react';

export interface AuthContextProps {
    children: React.ReactNode;
    session: Session;
}

export default function AuthContext({ children, session }: AuthContextProps) {
    return (
        <SessionProvider session={session} refetchInterval={5 * 60} refetchOnWindowFocus={false}>
            {children}
        </SessionProvider>
    )
}

```

**Comment by ankurk91** [comment]
80% traffic on our website is coming from this package, using v5.0.0-beta.29	

![Image](https://github.com/user-attachments/assets/7138c5e5-5299-4195-95c9-972c37e1b547)

**Comment by lunarW1TCH** [comment]
[Above i mentioned a non-perfect workaround.](#issuecomment-2407550784)

The perfect workaround was to switch fully to [better-auth](https://www.better-auth.com/ )
Code isn't that much different and the library **actually works the way it should**.

**Comment by Aman-456** [comment]
Hey,

This issue is still happening in **Auth.js v5** — even with `refetchOnWindowFocus={false}` and `refetchInterval={0}`, `SessionProvider` + `useSession()` often fires **two identical** `/api/auth/session` requests on mount due to the internal BroadcastChannel self-echo.

I built a small zero-dependency package that deduplicates these at the `fetch` level (short TTL coalescing so it doesn't affect cross-tab sync):

→ https://www.npmjs.com/package/nextauth-session-dedupe

**Usage:**
```tsx
import "nextauth-session-dedupe/auto";   // add this in your providers file
```

**Important note on Axios**:  
This package only affects native `fetch` calls. If you're using Axios with its default XHR adapter (most common), it won't deduplicate those requests. In that case, you'd need to either:
- Configure Axios to use the `fetch` adapter (`adapter: 'fetch'`), or
- Handle deduplication manually for your Axios instance.

Would love feedback if this helps the v5 cases!

**Comment by Aman-456** [comment]
Hey,

This issue is still happening in Auth.js v5 — even with `refetchOnWindowFocus={false}` and `refetchInterval={0}`, `SessionProvider` + `useSession()` often fires two identical `/api/auth/session` requests on mount due to the internal `BroadcastChannel` self-echo.

I built a small zero-dependency package that deduplicates these at the fetch level (short TTL coalescing so it doesn’t affect cross‑tab sync):

→ https://www.npmjs.com/package/fetch-coalescer  
→ https://github.com/Aman-456/fetch-coalescer

Usage:

```ts
import "fetch-coalescer/auto"; // add this in your providers file
```

Important note on Axios:
This package only affects native `fetch` calls. If you’re using Axios with its default XHR adapter (most common), it won’t deduplicate those requests. In that case, you’d need to either:

- Configure Axios to use a fetch adapter, or
- Handle deduplication manually for your Axios instance.

Would love feedback if this helps the v5 cases!
